drivers: ieee802154: nrf5: sleep if idle when `RxOnWhenIdle=0`

edmont opened the pull request:
When RxOnWhenIdle is set to False, turn the radio off
if no operation is ongoing in order to save power.

pdgendt reviewed (dismissed):
I think the `ARG_UNUSED(aInstance);` line can be removed?

pdgendt commented on modules/openthread/platform/radio.c:
```diff
@@ -920,11 +920,13 @@ void otPlatRadioSetRxOnWhenIdle(otInstance *aInstance, bool aRxOnWhenIdle)
 	radio_api->configure(radio_dev, IEEE802154_CONFIG_RX_ON_WHEN_IDLE, &config);
+
+	if (!aRxOnWhenIdle) {
+		ARG_UNUSED(otPlatRadioSleep(aInstance));
```
Are you sure you are idle when calling this?

abtink commented on modules/openthread/platform/radio.c:
```diff
@@ -920,11 +920,13 @@ void otPlatRadioSetRxOnWhenIdle(otInstance *aInstance, bool aRxOnWhenIdle)
+
+	if (!aRxOnWhenIdle) {
+		ARG_UNUSED(otPlatRadioSleep(aInstance));
+	}
```
Copying comment from https://github.com/nrfconnect/sdk-zephyr/pull/1751/files#r1619422853.

Thanks @edmont.

The "RxOnWhenIdle" flag can be changed by various triggers within the OT stack. For instance, the `otThreadSetLinkMode()` API, called by higher-level code controlling the OT stack, can switch the device mode, making it sleepy or Rx-on at any point in time.

Therefore, we should consider that `otPlatRadioSetRxOnWhenIdle()` could also be invoked at any time, while the radio may be in different states. For example, the radio could be in the middle of frame transmission or performing an energy scan.

Would calling `otPlatRadioSleep()` in all cases be appropriate? For instance, if the radio is mid-transmission, the OT stack expects a "send done" callback after a successful `otPlatRadioTransmit()`. Will this behave correctly after calling `otPlatRadioSleep()` in such a case?

edmont replied:
Please, see my comment above. I'll move the logic to platform shim layer for now.

fgrandel commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -1002,10 +1002,14 @@ static int nrf5_configure(const struct device *dev,
-		break;
+
+		if (config->rx_on_when_idle == false) {
+			ARG_UNUSED(nrf_802154_sleep_if_idle());
```
```suggestion
			(void)nrf_802154_sleep_if_idle();
```
I believe it's quite uncommon to use `ARG_UNUSED` in this case as it usually documents unused function arguments.

pdgendt commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -1002,10 +1002,14 @@ static int nrf5_configure(const struct device *dev,
 #endif /* CONFIG_IEEE802154_NRF5_MULTIPLE_CCA */
 
-	case IEEE802154_CONFIG_RX_ON_WHEN_IDLE:
+	case IEEE802154_CONFIG_RX_ON_WHEN_IDLE: {
```
The additional code block isn't required.

pdgendt reviewed (dismissed):
I think the commit/PR subject has changed drastically.

edmont committed:
drivers: ieee802154: nrf5: sleep if idle when RxOnWhenIdle=0

When RxOnWhenIdle is set to False, turn the radio off
if no operation is ongoing in order to save power.

Signed-off-by: Eduardo Montoya <eduardo.montoya@nordicsemi.no>
